### ISSUE TITLE
[Label Redesign] Update age_load

### ISSUE THREAD

**Issue by rafsun42** [apache/age#1090]

**Comment by Zainab-Saad** [comment]
Assign me please

**Comment by mohayu22** [comment]
Assign this to me

**Comment by rafsun42** [comment]
@Zainab-Saad Does sorting the vertex tables and edge csv file help?

**Comment by Zainab-Saad** [comment]
@rafsun42 Which field of the vertex table or edges csv should we sort? 
The primary performance issue is that I need to bring vertex tables into memory before creating the edges (to get the id of the vertex). This is the point that needs optimization.
I am not sure how sorting would help reduce the run time in this case.

**Comment by rafsun42** [comment]
@Zainab-Saad 
I was thinking about an algorithm that merges two sorted array. In our case, the vertex table and edge csv file. And, we are not actually merging them. We are just scanning the vertex table to get the new id. 

Sort the vertex table by `id`. Sort the edge csv file by `start_id` and `end_id`. You can set up a cursor to scan the vertex table row by row. As you read the edge csv file, you can get the next vertex from the cursor, instead of scanning the entire vertex table. Because these tables are sorted, you need to scan vertex table and csv file only once. 

This is just an idea. One drawback is sorting will cost time, but may be worth it for large csv file.

Another idea that Shoaib suggested is to use the IDs in the csv file as actual ID (instead of getting it from the sequence), while importing vertices. Once imported, set the sequence to reflect the max ID in CSV + 1.

Both of these are just ideas, you may need to think about their correctness and any potential edge cases.

**Comment by Zainab-Saad** [comment]
@rafsun42 I have tried analyzing the two ideas

Regarding @muhammadshoaib 's idea, we can have the same performance as before (as we wont be scanning any table) but there is a loop hole what if two different csv files have the same `id` column value, this will cause clashes in vertex id. If we can find a way to go about this issue, then this idea has potential because previously too the function depended on `id` column as it used this as the `entry id` for generating `graphid`. Another issue could be that we might waste sequence numbers (I am not sure if this would be a big issue as the max sequence number is `9223372036854775807` )

Regarding your idea, the only issue might be the vertex table scan (which will take the most time even if done once)

**Comment by rafsun42** [comment]
@Zainab-Saad 

Feel free to think about any addition you would like to make to this idea to go around the issue. I and Shoaib will discuss it after.

**Comment by Zainab-Saad** [comment]
@rafsun42 
When creating vertices from the csv, we can cater duplicate ids (for example by adding timestamps to id) but that will break the edge csv because edge csv holds the ids that were originally present in the vertex csv
One approach could be to not allow such id duplication altogether.

**Comment by rafsun42** [comment]
@Zainab-Saad 

Let's say there is no duplication in vertex csv files. And, we are planning to use the csv IDs as actual ID.

1. `load_vertex_from_csv (vertex1)`
2. `CREATE (x:vertex1)
3. `load_vertex_from_csv (vertex2)`
4. `load_edge_from_csv(edge1)`

Would the `CREATE` clause break this solution?

**Comment by Zainab-Saad** [comment]
@rafsun42 
After step 1, if we reset the sequence starting number to `max CSV ID + 1` and then create a vertex as in step 2, there would not be an issue till this point. But as we move to the step 3, if the vertex2 csv has an id that we have already used in step 2 in `CREATE(x: vertex2)` then this will cause id clash and a possible break.
PS: I am considering here the id clash from the CREATE clause and the vertex2 csv not across multiple vertex csvs (which we are ignoring for now as per your comment)

**Comment by rafsun42** [comment]
@Zainab-Saad 
Does it solve the problem if `actual ID = next available ID in the sequence + ID in csv`?

**Comment by Zainab-Saad** [comment]
@rafsun42 
That solves the id clash but will break the edge csv (as it still has the previous ids)

**Comment by panosfol** [comment]
assign me please

**Comment by Zainab-Saad** [comment]
@rafsun42 
I have created a [PR](https://github.com/rafsun42/age/pull/23) that updates age_load and the relevant regression tests (tests pass).
Once this is approved, we can start working on supporting multiple labels with age_load.
